결제 승인 완료 이벤트 발행 및 처리 카프카 프로듀서 컨슈머로 대체

wanniDev opened the pull request:
## 카프카를 활용하는 이유
스프링에서 제공하는 Event를 활용해도 기존 결제 서비스를 지연시키지 않고 승인 완료 여부를 db에 저장하는 이벤트를 비동기로 처리 가능합니다.
하지만, 장애 지점별 결제 이력 손실을 최소화하기 위해서 최대한 다양한 지점에 로그를 기록할 필요가 있습니다.

이벤트 처리 및 이벤트 로그를 기록하는 기능은 스프링 프레임워크 만으로도 처리가 가능하지만, 비동기 이벤트 처리와 동시에 이벤트 발생 순서에 따라 로그를 기록하는 기능은 매우 복잡하고 장애 발생 시, 대응하기도 까다롭습니다.

그래서 카프카를 활용하여, 이벤트를 비동기로 처리하여 결제 서비스의 불필요한 지연 시간을 줄여, 사용자 경험을 침해하지 않고 로그 시스템 구현 복잡도를 줄이고, 장애가 발생하지 않는 안정적인 구조를 유지하고자 합니다. 해당 pr에서 진행한 작업은 다음과 같습니다.

## 카프카 클러스터 구조
![payment_kafka drawio](https://github.com/f-lab-edu/payment-lab/assets/81374655/1fbbd8ba-b3ac-452e-8af8-12e7a0d208f2)

- 아직은 카프카 클러스터를 운영할 단계는 아니기 때문에 도커 컴포즈를 활용하여, 단일 브로커, 프로듀서, 컨슈머로 매우 간단하게 구성하였습니다.
- 토픽 네이밍은 [kafka topics naming conventions](https://medium.com/@kiranprabhu/kafka-topic-naming-conventions-best-practices-6b6b332769a3)를 참고하여, `<환경>.<도메인>.<서비스기능>.<메시지-타입>`으로 명명했습니다.
- 불필요한 layer을 구성하여 코드 복잡도를 높이는 상황을 피하기 위해, 결제 승인 이력 이벤트 발행 및 처리는 `kafkaTemplate`와 `@KafkaListener`로 대체하였습니다.
- 카프카 컨슈머 그룹과 토픽 선정은 평문보다는 yml 프로퍼티 경로로 대체하여 재사용성을 높이고자 하였습니다.
- 약간의 성능 손실이 있지만, 이벤트 메시지의 직렬화와 역직렬화 과정을 직관적이고 유연하게 사용하기 위해 `StringDeserializer`를 활용했습니다.

wanniDev committed:
Feat: 결제 완료 이벤트 처리에 필요한 카프카 도커 컴포즈 작성

아직 운영 단계는 아니기 때문에 카프카 시스템 설정에 대한 부분은 추후 학습하며 보완하고
지금은 기본적인 설정을 기반으로 producer, consumer 간의 이벤트 서빙을 제대로하는 것에 집중

wanniDev committed:
Merge branch 'f-lab-edu:main' into main

wanniDev committed:
Feat: 이벤트 브로커 구성 및 사용을 위해 카프카 의존성 추가

wanniDev committed:
Feat: 결제 승인 이벤트 발행을 담당하는 카프카 프로듀서 설정 적용

주키퍼 및 브로커 서버와 파티션 설정은 설정 내용을 숨기고 어차피 카프카에 장애가 생기면 브로커 서버 및 주키퍼를
시스템상에서 진단을 해야하므로 카프카 cli로 구성 하였다.

카프카 토픽의 네이밍은 기존 컨벤션을 참고하여 <환경>.<도메인>.<서비스기능>.<메시지-타입>으로 구성했다.

wanniDev committed:
Feat: 기존의 결제 트랜잭션 승인 발행 기능을 카프카 프로듀서로 대체

- 기존에 발행되는 이벤트의 처리는 오직 트랜잭션 로그 백업 행위에만 적용
- 새로 적용된 카프카는 결제 승인 이벤트 처리 및 해당 승인 이력 로그 백업을 담당한다.
- spring kafka에서 제공하는 kafkaTemplate은 기본적으로 메시지를 보내는 작업은 비동기로 수행한다.(ListenableFuture)
따라서, 해당 이벤트 처리에 이상이 생기더라도 토스페이의 기존 결제 승인 api 결과를 리턴하는 작업에는 지장을 주지 않는다.

wanniDev committed:
Feat: 결제 승인 이력 저장 이벤트를 처리하는 역할을 카프카 리스너에게 위임

기존에는 @EventListener를 별도로 구성하였으나, 카프카 클라이언트를 활용하는 시점에서는
굳이 @EventListener를 수행하는 컴포넌트를 따로 구성할 필요는 없을 거 같다.

그래서, 결제 승인 이력 저장 이벤트 처리는 @KafkaListener로 대체하였다.

카프카의 경우, 컨슈머의 성공여부에 상관없이 해당 이력을 로그로 기록을 수행한다.

그리고, 카프카의 프로듀서가 이벤트를 발행하는 시점에서는 이미 토스 결제 승인이 성공이된 이후이기 때문에
이벤트 처리 성공여부에 상관없이 결제는 성공처리되는 것이 타당하다.

wanniDev committed:
Refactor: 유연한 코드관리를 위해 @KafkaListener 애노테이션의 파라미터 값을 평문에서 yml 프로퍼티 경로로 수정

wanniDev committed:
Move: 다양한 모듈 전역으로 공유되는 카프카 컨슈머, 프로듀서 구성 설정 빈을 app 모듈로 이동

f-lab-troy commented on docker-compose-kafka-single-broker.yml:
```diff
@@ -0,0 +1,28 @@
+version: '2'
+
+services:
+  zookeeper:
```
zookeeper와 kafka의 데이터를 컨테이너 밖의 볼륨으로 mapping 해주셔서 따로 저장되게 해주시면 좋을 것 같습니다.
참고: https://stackoverflow.com/questions/59592518/kafka-broker-doesnt-find-cluster-id-and-creates-new-one-after-docker-restart

wanniDev committed:
Fix: 카프카 브로커 재가동시 발생하는 cluster id 이슈 대응

https://stackoverflow.com/questions/59592518/kafka-broker-doesnt-find-cluster-id-and-creates-new-one-after-docker-restart
